lib/rbi/rewriters/flatten_scopes.rb

```diff
@@ -0,0 +1,44 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RBI
+  module Rewriters
+    class FlattenScopes < Visitor
+      extend T::Sig
+
+      sig { override.params(node: T.nilable(Node)).void }
+      def visit(node)
+        return unless node
+
+        case node
+        when Tree
+          visit_all(node.nodes)
+
+          parent_tree = node.parent_tree
+          return unless parent_tree
+
+          node.nodes.dup.each do |child|
+            next unless child.is_a?(Class) || child.is_a?(Module)
```

Comment:
Why do we need to iterate over the child nodes of the tree?

Can't we have a `when` for those instead?

Something like this:

```rb
case node
when Class, Module
  visit_all(node.nodes)

  parent_tree = node.parent_tree
  return unless parent_tree

  parent_scope = node.parent_scope
  return unless parent_scope.is_a?(Class) || parent_scope.is_a?(Module)

  node.detach
  node.name = "#{parent_scope.name}::#{node.name}"

  parent_tree << node
when Tree
  visit_all(node.nodes)
end
```

Reply:
I tried this type of approach and couldn't get it to work. Where you have `parent_tree` you actually need "grandparent" (i.e. `node.parent_tree&.parent_tree`) since we're flattening by removing a node from its parent, adding the namespace to the name, and reattaching to the grandparent. Even then, I couldn't exactly figure out why (got a bit lost in the recursion) but I think it has to do with the order in which you're visiting and detaching/reattaching the nodes.

You're welcome to play around with it yourself and push a commit if you can get it working, but it's pretty terse as-is so I think it may be OK.